Aarch64 jit depthwise convolution kernels

nishith-fujitsu opened the pull request:
# Description
This commit expands ARM SVE support for forward and backward JIT SVE Depth-wise convolution in FP32 for, introducing compatibility with various vector lengths. The changes made are for implementing different ARM ISA.
Major code changes:

1. Added common files _jit_uni_dw_convolution.cpp, jit_uni_dw_convolution.hpp, jit_uni_dw_conv_kernel_utils.hpp, jit_uni_dw_conv_kernel_f32.cpp, jit_uni_dw_conv_kernel_f32.hpp_ to accommodate the extended ARM SVE ISA for depth-wise convolution operators.
2. Set data format tags according to the ARM ISA being used for forward and backward operator in JIT depth-wise convolution.
3. Replaced ldr, and str instructions for vector registers with ld1w and st1w to utilize predication.
4. Made changes to reducer.cpp to support Vector agnostic approach.
 

# Checklist

## General

- [✓]  Do all unit and benchdnn tests (`make test` and `make test_benchdnn_*`) pass locally for each commit? Yes
Test output is same with and without this commit.

```
99% tests passed, 2 tests failed out of 196
 
Total Test time (real) = 525.25 sec
 
The following tests FAILED:
        155 - test_graph_unit_dnnl_large_partition_usm_cpu (Failed)
        177 - test_benchdnn_modeC_graph_ci_cpu (Failed)
Errors while running CTest
Output from these tests are in: /home/nishith/oss/oneDNN/build/Testing/Temporary/LastTest.log
Use "--rerun-failed --output-on-failure" to re-run the failed cases verbosely.
make: *** [Makefile:71: test] Error 8
```

- [ ✓] Have you formatted the code using clang-format? Yes

abhijain1204fujitsu commented:
@vpirogov , Can you please share your feedback for the PR & Support us for merger

vpirogov commented:
@oneapi-src/onednn-cpu-aarch64, @snadampal, @kawakami-k, could you please help reviewing this one?

jondea commented on src/cpu/cpu_convolution_list.cpp:
```diff
@@ -138,12 +138,13 @@ const std::map<pk_dt_impl_key_t, std::vector<impl_list_item_t>> &impl_list_map()
             CPU_INSTANCE_AARCH64_ACL(acl_depthwise_convolution_fwd_t)
             CPU_INSTANCE_AARCH64_ACL(acl_indirect_gemm_convolution_fwd_t)
             CPU_INSTANCE_AARCH64_ACL(acl_gemm_convolution_fwd_t<f32>)
+            CPU_INSTANCE_AARCH64(jit_uni_dw_convolution_fwd_t<sve_256,data_type::f32>)
```
Could you move this above ACL in the list please?

I've done some benchmarking at the oneDNN and PyTorch levels, and it looks like for Neoverse-V1, this implementation is better than the existing ACL implementation. This is really great work, thank you!

abhijain1204fujitsu replied:
Hi @jondea , will check and do the needful

nishith-fujitsu replied:
Hi @jondea, thank you for reviewing my PR. I have done the needful changes.

jondea commented on src/cpu/aarch64/jit_uni_dw_convolution.hpp:
```diff
@@ -96,9 +96,11 @@ struct jit_uni_dw_convolution_fwd_t : public primitive_t {
     std::unique_ptr<jit_uni_dw_conv_fwd_kernel<isa, src_type>> kernel_;
 };
-
+// To remove after resolving dependency on 1x1
```
I don't understand this comment, sorry. Could you elaborate please?

nishith-fujitsu replied:
Hi @jondea earlier this variable is used to call template code in dispatch list for the kernel. But now I have removed this variable from the code from cpu_convolution_list. But same variable template is still used to call the DW kernel in JIT_1x1_kernel files. We will remove this in future release.

jondea commented on src/cpu/aarch64/cpu_reducer.cpp:
```diff
@@ -217,7 +217,11 @@ struct reducer_2d_driver_f_s_32_t : public reducer_2d_driver_t<data_type, isa> {
         Label loop_x_label[nbranches + 1];
 
-        this->ptrue(preg_all.b);
+        switch (isa) {
```
I don't have a good understanding of xbyak for SVE, sorry, but do you know why this is necessary? Why can't we just use plain old ptrue?

nishith-fujitsu replied:
Hi @jondea, we edited this code to support multiple ISA vector lengths for our kernels.

jondea replied:
I understand that was the general idea, but why is `VL32` and `VL64` needed, rather than the default `ALL`? Was it so that you could test the 256 kernels on a machine with 512 vector length?

nishith-fujitsu replied:
yes, you got it right it right. It enables sve_256 kernels on 512 VL machine.

jondea reviewed (changes requested):
This is really great work, thank you!

My only important comment is that I think it is better than the current ACL depthwise implementation and it should go ahead of it in the dispatch list.

I have also done some testing with this on a C7g with benchdnn and with PyTorch, and I didn't hit any issues.

jondea reviewed (changes requested):
Code generally looks good, but can you remove the merge commit from the history by rebasing onto main. https://github.com/oneapi-src/oneDNN/blob/main/CONTRIBUTING.md#code-contribution-guidelines

abhijain1204fujitsu commented:
@vpirogov , @jondea the requested changes has been completed, kindly check and support to merge the PR

abhijain1204fujitsu commented:
@vpirogov kindly support to merge the changes
In case there is any more feedback kindly let us know. 
Thank you !

mgouicem reviewed (changes requested):
I believe a missing closing brace should make that code break the build on ARM platforms.

nishith-fujitsu commented on src/cpu/aarch64/jit_uni_dw_convolution.hpp:
```diff
@@ -95,180 +95,206 @@ struct jit_uni_dw_convolution_fwd_t : public primitive_t {
-            auto scratchpad = scratchpad_registry().registrar();
-            jit_uni_dw_conv_bwd_data_kernel<isa,
-                    diff_dst_type>::init_scratchpad(scratchpad, jcp_);
+    using jit_sve_512_dw_convolution_fwd_t
```
Hi @mgouicem, thank you for reviewing my code. It was missed in latest changes. I have made the suggested changes.

mgouicem reviewed (changes requested):
please remove the merge commit. Instead rebase your branch on top of the main branch.

nishith-fujitsu commented:
> please remove the merge commit. Instead rebase your branch on top of the main branch.

Hi @mgouicem, removed merge commits and rebased to latest main branch.

abhijain1204fujitsu commented:
@vpirogov , @mgouicem please support to merge the PR
